File: restaurant-management-backend/Utility/ApiResponse.cs
﻿namespace restaurant_management_backend.Utility
{
    public class ApiResponse<T>
    {
        public bool Success { get; set; }
        public string? Message { get; set; }
        public int StatusCode { get; set; }
        public T? Data { get; set; }

        public static ApiResponse<T> SuccessResponse(T? data, string? message = null, int statusCode = 200)
        {
            return new ApiResponse<T>
            {
                Success = true,
                Message = message,
                StatusCode = statusCode,
                Data = data
            };
        }

        public static ApiResponse<T> FailureResponse(string message, int statusCode = 400)
        {
            return new ApiResponse<T>
            {
                Success = false,
                Message = message,
                StatusCode = statusCode,
                //Data = default
            };
        }

    }

}

File: restaurant-management-frontend/src/app/core/models/api-response.ts
export class ApiResponse<T> {
  success: boolean;
  message: string;
  statusCode: number;
  data?: T | null;

  private constructor(
    success: boolean,
    message: string,
    statusCode: number,
    data: T | null
  ) {
    this.success = success;
    this.message = message;
    this.statusCode = statusCode;
    this.data = data ?? null;
  }

  static successResponse<T>(data?: T | null, message: string='', statusCode: number = 200): ApiResponse<T> {
    return new ApiResponse<T>(true, message, statusCode, data ?? null);
  }

  static failureResponse<T>(message: string, statusCode: number = 400): ApiResponse<T> {
    return new ApiResponse<T>(false, message, statusCode, null);
  }
}
